ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java

```diff
@@ -2436,6 +2438,578 @@ static void undeleteRecoveryTest(Port targetPort, MockCluster cluster, SSLConfig
+          properties.getAccountId(), properties.getContainerId(), partitionId, false, BlobId.BlobDataType.DATACHUNK);
+      BlobId blobId14 = new BlobId(blobIdVersion, BlobId.BlobIdType.NATIVE, clusterMap.getLocalDatacenterId(),
+          properties.getAccountId(), properties.getContainerId(), partitionId, false, BlobId.BlobDataType.DATACHUNK);
+      BlobId blobId15 = new BlobId(blobIdVersion, BlobId.BlobIdType.NATIVE, clusterMap.getLocalDatacenterId(),
+          properties.getAccountId(), properties.getContainerId(), partitionId, false, BlobId.BlobDataType.DATACHUNK);
+
+      Port sourcePort = new Port(sourceDataNode.getHttp2Port(), PortType.HTTP2);
+      Port targetPort = new Port(targetDataNode.getHttp2Port(), PortType.HTTP2);
+      Port comparisonPort = new Port(comparisonDataNode.getHttp2Port(), PortType.HTTP2);
+      ConnectedChannel sourceChannel =
+          getBlockingChannelBasedOnPortType(sourcePort, "localhost", null, clientSSLConfig);
+      ConnectedChannel targetChannel =
+          getBlockingChannelBasedOnPortType(targetPort, "localhost", null, clientSSLConfig);
+      ConnectedChannel comparisonChannel =
+          getBlockingChannelBasedOnPortType(comparisonPort, "localhost", null, clientSSLConfig);
+      sourceChannel.connect();
+      targetChannel.connect();
+      comparisonChannel.connect();
+
+      /*
+       * Stage 1: Test the correctness of the ReplicateBlob. Replicate blobs from sourceDataNode to the targetDataNode
+       * On the sourceDataNode and the targetDataNode, the blob is under different states.
+       */
+      // 1. On the sourceDataNode, putBlob. On the targetDataNode, blob doesn't exist.
+      // ReplicateBlob replicates the blob to the target.
+      BlobId blobId = blobId1;
+      PutRequest putRequest = new PutRequest(1, "client1", blobId, propertiesWithTtl, ByteBuffer.wrap(userMetadata),
+          Unpooled.wrappedBuffer(data), properties.getBlobSize(), BlobType.DataBlob,
+          testEncryption ? ByteBuffer.wrap(encryptionKey) : null);
+      putBlob(putRequest, sourceChannel, ServerErrorCode.No_Error);
+      replicateBlob(targetChannel, blobId, sourceDataNode, ServerErrorCode.No_Error);
+      getBlobAndVerify(blobId, targetChannel, ServerErrorCode.No_Error, propertiesWithTtl, userMetadata, data,
+          encryptionKey, clusterMap, blobIdFactory, testEncryption);
+
+      // 2. On the sourceDataNode, putBlob. On the targetDataNode, the blob already exists.
+      // ReplicateBlob to both the sourceDataNode and the targetDataNode. All succeed.
+      blobId = blobId2;
+      putRequest = new PutRequest(1, "client1", blobId, propertiesWithTtl, ByteBuffer.wrap(userMetadata),
+          Unpooled.wrappedBuffer(data), properties.getBlobSize(), BlobType.DataBlob,
+          testEncryption ? ByteBuffer.wrap(encryptionKey) : null);
+      putBlob(putRequest, sourceChannel, ServerErrorCode.No_Error);
+      replicateBlob(sourceChannel, blobId, sourceDataNode, ServerErrorCode.No_Error);
```

Comment:
this is a great test. but in AmbryRequests implementation, we should probably check if the source host  is the same host as localhost, if so we should just return true right away.

Reply:
Yes, plan to do that. Will file a separate PR for it.